### PR TITLE
[Fundamental] Enable recording playback for SDK/CLI Azure end-to-end tests

### PR DESCRIPTION
**Pull request by zhengfeiwang**
# Description

This PR targets to introduce recording capability for SDK/CLI Azure end-to-end tests. A new CI pipeline "promptflow-sdk-cli-azure-e2e-test" will be added, tests under `sdk_cli_azure_test` will be executed, and there will be no real network request and all requests will be mocked leveraging [VCR.py](https://pypi.org/project/vcrpy/).

This PR does skip some tests due to various reasons, in summary, CLI/telemetry/flow in pipeline tests are skipped; and some run related tests are skipped and there are still some works to enable - might do it in future PR.

This is the first step we unblock 3P contributor as forked PR cannot retrieve GitHub repo secrets, so recording can unblock them as it does not require real network request, nor real secrets. The next step is to make current SDK/CLI CI not required for PR and run daily, meanwhile, the new CI will become the required one.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/tests/sdk_cli_azure_test/_fake_credentials.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class FakeTokenCredential:
+    """Refer from Azure SDK for Python repository.
+
+    https://github.com/Azure/azure-sdk-for-python/blob/main/tools/azure-sdk-tools/devtools_testutils/fake_credentials.py
+    """
+
+    def __init__(self):
+        self.token = AccessToken("YOU SHALL NOT PASS", 0)
```
Comment (crazygao) [inline src/promptflow/tests/sdk_cli_azure_test/_fake_credentials.py:15]: :D

File: src/promptflow/setup.py [hunk 2]
```diff
@@ -46,6 +46,7 @@
     "waitress>=2.1.2,<3.0.0",  # used to serve local service
     "opencensus-ext-azure<2.0.0",  # configure opencensus to send telemetry to azure monitor
     "ruamel.yaml>=0.17.35,<0.18.0",  # used to generate connection templates with preserved comments
+    "vcrpy>=5.1.0,<6.0.0",  # used to record and replay http requests for pfazure tests
```
Comment (0mza987) [inline src/promptflow/setup.py:49]: Think this should be in `dev_requirements.txt`, just like `pytest` dependencies?

File: src/promptflow/tests/sdk_cli_azure_test/_recording_base.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        test_file_recording_dir = (recording_dir / test_file_name).resolve()
+        test_file_recording_dir.mkdir(parents=True, exist_ok=True)
+        test_class_name = self.__class__.__name__
+        recording_filename = f"{test_class_name}_{method_name}.yaml"
+        self.recording_file = (test_file_recording_dir / recording_filename).resolve()
+
+        if self.is_live and not is_live_and_not_recording() and self.recording_file.is_file():
+            self.recording_file.unlink()
+
+    def setUp(self) -> None:
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_azure_test/_recording_base.py:81]: why camel here

File: src/promptflow/tests/sdk_cli_azure_test/conftest.py [hunk 4]
```diff
@@ -126,3 +158,9 @@ def ml_client_canary(
         workspace_name="promptflow-canary-dev",
         cloud="AzureCloud",
     )
+
+
+@pytest.fixture(autouse=True)
+def mock_appinsights_log_handler(mocker: MockFixture) -> None:
+    dummy_logger = logging.getLogger("dummy")
+    mocker.patch("promptflow._telemetry.telemetry.get_telemetry_logger", return_value=dummy_logger)
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_azure_test/conftest.py:166]: Why add this? I thought test logs won't get into telemetry in the first place?

File: src/promptflow/tests/sdk_cli_azure_test/_fake_credentials.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+from azure.core.credentials import AccessToken
+
+
+class FakeTokenCredential:
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_azure_test/_fake_credentials.py:8]: Weird to put these files under `sdk_cli_azure_test`, can we put them in `tests/test_utilities` instead?

File: src/promptflow/tests/sdk_cli_azure_test/conftest.py [hunk 6]
```diff
@@ -53,9 +84,10 @@ def remote_client_int() -> PFClient:
     return PFClient(ml_client=client)
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_azure_test/conftest.py:84]: Seems we have many other clients `rempte_client_int`, `ml_client_canary` etc. Should we cover all of them?

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_connection_operations.py [hunk 7]
```diff
@@ -1,25 +1,30 @@
...
+from .._recording_utils import fixture_provider
 
-@pytest.fixture
-def connection_ops(ml_client):
-    from promptflow.azure import PFClient
 
-    pf = PFClient(ml_client=ml_client)
-    yield pf._connections
+@pytest.fixture(scope="class")
+def connection_ops(request: pytest.FixtureRequest, pf: PFClient) -> ConnectionOperations:
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_azure_test/e2etests/test_connection_operations.py:20]: seems this fixture appears in multiple files? How about put them in the conftest

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py [hunk 8]
```diff
@@ -30,9 +33,11 @@
 
 
 # TODO(2528577): we should run these test with recording mode.
+@pytest.mark.usefixtures("pf", "runtime")
 @pytest.mark.e2etest
-class TestFlowRun:
-    def test_run_bulk(self, remote_client, pf, runtime):
+class TestFlowRun(PFAzureIntegrationTestCase):
+    @fixture_provider
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py:39]: why need this provider? Our tests should all be pytest test case?

File: src/promptflow/tests/sdk_cli_azure_test/_recording_base.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+    ]
+
+    def __init__(self, method_name: str) -> None:
+        super(PFAzureIntegrationTestCase, self).__init__(method_name)
+
+        self.recording_processors = self._get_recording_processors()
+        self.playback_processors = self._get_playback_processors()
+
+        test_file_path = Path(inspect.getfile(self.__class__)).resolve()
+        recording_dir = (test_file_path.parent / "recordings").resolve()
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_azure_test/_recording_base.py:57]: Shall we centralize all recordings to one folder instead of under lots of test folders?

File: src/promptflow/setup.py (hunk 2 above)
Comment (zhengfeiwang) [reply src/promptflow/setup.py:49]: good catch, should be moved

File: src/promptflow/tests/sdk_cli_azure_test/_recording_base.py (hunk 3 above)
Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_azure_test/_recording_base.py]: unit test class method; have removed that for pure pytest experience

File: src/promptflow/tests/sdk_cli_azure_test/_fake_credentials.py (hunk 5 above)
Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_azure_test/_fake_credentials.py]: that should only be needed inside pfazure test, so in new implementation, still put it under `sdk_cli_azure_test`

File: src/promptflow/tests/sdk_cli_azure_test/conftest.py (hunk 6 above)
Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_azure_test/conftest.py]: updated, no need for that change any more

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py (hunk 8 above)
Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py]: no need to this decorator any more

File: src/promptflow/tests/sdk_cli_azure_test/_recording_base.py (hunk 9 above)
Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_azure_test/_recording_base.py]: good idea, have put them under `tests/test_configs/recording` in new impl

File: src/promptflow/tests/sdk_cli_azure_test/conftest.py (hunk 4 above)
Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_azure_test/conftest.py:154]: this fixture targets to mock the init operation, which will generate some requests with sensitive data

File: .github/workflows/promptflow-sdk-cli-azure-e2e-test.yml [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        with:
+          pythonVersion: ${{ steps.display_env.outputs.pyVersion }}
+
+      - name: Build wheel
+        uses: "./.github/actions/step_sdk_setup"
+        with:
+          setupType: ${{ env.packageSetupType }}
+          scriptPath: ${{ env.testWorkingDirectory }}
+
+      # - name: Generate Configs
```
Comment (crazygao) [inline .github/workflows/promptflow-sdk-cli-azure-e2e-test.yml:55]: Delete this part

**Review by crazygao — DISMISSED** [review]

File: .github/workflows/promptflow-sdk-cli-azure-e2e-test.yml (hunk 10 above)
Comment (zhengfeiwang) [reply .github/workflows/promptflow-sdk-cli-azure-e2e-test.yml]: removed

File: scripts/building/dev_requirements.txt [hunk 11]
```diff
@@ -9,3 +9,8 @@ pytest-sugar
 pytest-timeout
 azure-keyvault
 azure-identity
+# record and replay http requests for pfazure tests
```
Comment (zhengfeiwang) [inline scripts/building/dev_requirements.txt:12]: can we use `src/promptflow/dev_requirements.txt`? otherwise we might need double updates for dev reqs.

Comment (0mza987) [reply scripts/building/dev_requirements.txt:12]: Think we can use that, that's the dev requirement file what my original comment implied.

Comment (0mza987) [reply scripts/building/dev_requirements.txt:12]: Why need this duplicate since `src/promptflow/dev_requirements.txt` already has the deps

Comment (zhengfeiwang) [reply scripts/building/dev_requirements.txt:12]: @crazygao ping on this

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_flow_serve.py [hunk 12]
```diff
@@ -2,7 +2,10 @@
 
 import pytest
 
+from ..recording_utilities import is_live
 
+
+@pytest.mark.skipif(condition=not is_live(), reason="serving tests, only run in live mode.")
```
Comment (0mza987) [inline src/promptflow/tests/sdk_cli_azure_test/e2etests/test_flow_serve.py:8]: Seems several types of tests are skipped if not live mode, why is that necessary?

Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_azure_test/e2etests/test_flow_serve.py:8]: currently default mode is live but not recording, so the experience is same as before; for those skipped tests, I will open another PR to enable as much as possible in the near future.

**Review by 0mza987 — APPROVED** [review]

File: src/promptflow/tests/test_configs/recordings/test_run_operations_TestFlowRun_test_run_with_connection_overwrite.yaml [hunk 13]
```diff
@@ -0,0 +1,857 @@
...
+      Content-Length:
+      - '0'
+      User-Agent:
+      - promptflow-sdk/0.0.1 azure-ai-ml/1.10.0 azsdk-python-mgmt-machinelearningservices/0.1.0
+        Python/3.10.13 (Windows-10-10.0.22621-SP0)
+    method: POST
+    uri: https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/00000/providers/Microsoft.MachineLearningServices/workspaces/00000/datastores/workspaceblobstore/listSecrets
+  response:
+    body:
+      string: '{"secretsType": "AccountKey", "key": "dGhpcyBpcyBmYWtlIGtleQ=="}'
```
Comment (wangchao1230) [inline src/promptflow/tests/test_configs/recordings/test_run_operations_TestFlowRun_test_run_with_connection_overwrite.yaml:168]: Is this sensitive?

File: src/promptflow/tests/sdk_cli_azure_test/recording_utilities/bases.py [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+    def _apply_replacement_for_recordings(self) -> None:
+        for i in range(len(self.cassette.data)):
+            req, resp = self.cassette.data[i]
+            req = self.variable_recorder.sanitize_request(req)
+            resp = self.variable_recorder.sanitize_response(resp)
+            self.cassette.data[i] = (req, resp)
+        return
+
+
+class PFAzureRunIntegrationTestRecording(PFAzureIntegrationTestRecording):
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_azure_test/recording_utilities/bases.py:161]: may add some comments on why we need different class for Run case

File: src/promptflow/tests/test_configs/recordings/test_arm_connection_operations_TestArmConnectionOperations_test_get_connection.yaml [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+  response:
+    body:
+      string: '{"tags": null, "location": null, "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/00000/providers/Microsoft.MachineLearningServices/workspaces/00000/connections/azure_open_ai_connection",
+        "name": "azure_open_ai_connection", "type": "Microsoft.MachineLearningServices/workspaces/connections",
+        "properties": {"authType": "ApiKey", "credentials": {"key": "_"}, "category":
+        "AzureOpenAI", "expiryTime": null, "target": "_", "createdByWorkspaceArmId":
+        null, "isSharedToAll": false, "sharedUserList": [], "metadata": {"azureml.flow.connection_type":
+        "AzureOpenAI", "azureml.flow.module": "promptflow.connections", "ApiType":
+        "azure", "ApiVersion": "2023-07-01-preview", "ResourceId": null}}, "systemData":
+        {"createdAt": "2023-08-22T10:15:34.5762053Z", "createdBy": "yigao@microsoft.com",
```
Comment (wangchao1230) [inline src/promptflow/tests/test_configs/recordings/test_arm_connection_operations_TestArmConnectionOperations_test_get_connection.yaml:122]: maybe we should sanitize emails

File: src/promptflow/tests/test_configs/recordings/test_cli_with_azure_TestCliWithAzure_test_basic_flow_run_bulk_without_env.yaml [hunk 16]
```diff
@@ -0,0 +1,764 @@
...
+      Content-Length:
+      - '0'
+      User-Agent:
+      - promptflow-sdk/0.0.1 azure-ai-ml/1.10.0 azsdk-python-mgmt-machinelearningservices/0.1.0
+        Python/3.10.13 (Windows-10-10.0.22621-SP0)
+    method: POST
+    uri: https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/00000/providers/Microsoft.MachineLearningServices/workspaces/00000/datastores/workspaceblobstore/listSecrets
+  response:
+    body:
+      string: '{"secretsType": "AccountKey", "key": "dGhpcyBpcyBmYWtlIGtleQ=="}'
```
Comment (wangchao1230) [inline src/promptflow/tests/test_configs/recordings/test_cli_with_azure_TestCliWithAzure_test_basic_flow_run_bulk_without_env.yaml:72]: Key?

File: src/promptflow/tests/sdk_cli_azure_test/conftest.py [hunk 17]
```diff
@@ -38,45 +61,51 @@ def ml_client(
...
+    if not is_live():
+        yield get_pf_client_for_playback()
+    else:
+        # enable telemetry for CI
+        with environment_variable_overwrite(TELEMETRY_ENABLED, "true"):
+            yield PFClient(
+                credential=get_cred(),
+                subscription_id="96aede12-2f73-41cb-b983-6d11a904839b",
+                resource_group_name="promptflow",
+                workspace_name="promptflow-eastus",
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_azure_test/conftest.py:75]: looks like we still have some test using eastus not canary

**Review by wangchao1230 — APPROVED** [review]

File: src/promptflow/tests/sdk_cli_azure_test/conftest.py (hunk 17 above)
Comment (zhengfeiwang) [reply src/promptflow/tests/sdk_cli_azure_test/conftest.py:75]: offline synced, currently our tests are still in canary region, will apply a switch later

File: src/promptflow/tests/test_configs/recordings/test_arm_connection_operations_TestArmConnectionOperations_test_get_connection.yaml (hunk 15 above)
Comment (zhengfeiwang) [reply src/promptflow/tests/test_configs/recordings/test_arm_connection_operations_TestArmConnectionOperations_test_get_connection.yaml:122]: opened a task to track this, will do it in next PR.

File: scripts/building/dev_requirements.txt (hunk 11 above)
Comment (zhengfeiwang) [reply scripts/building/dev_requirements.txt:12]: will unify them in future pr.
